##__1__. Подберите скорость обучения (alpha) и количество итераций (до совпадения ответов с результатами МНК):

In [1]:
import numpy as np

X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],   # для умножения на intercept
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]]) # стаж репетитора

y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60] # средний балл ЕГЭ (целевая переменная)

In [2]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

def calc_mae(y, y_pred):
    err = np.mean(np.abs(y - y_pred))
    return err

In [96]:
def model(X, alpha, weights, iter_count):
  n = X.shape[1]
  a = alpha
  w = np.array(weights)
  y_pred = 0
  w_needed = np.array([47.23214286,  3.91071429])
  for i in range(iter_count):
      y_pred = np.dot(w, X)
      err = calc_mse(y, y_pred)          
      for j in range(w.shape[0]):             
        w[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))
      if np.allclose(w, w_needed,rtol=1e-11) and err == 45.937499999999986:
          print(i, w, err)
          # МНК: array([47.23214286,  3.91071429]) 45.937499999999986  

## В коде выше возникла проблема со сравнением массивов (точность после запятой указана, но np.allclose всё равно возвращает True при 47.23214285. Другие методы сравнения вообще не работают. Пришлось глазами искать значение. Буду признателен, если подскажете: как сравнивать такие массивы?

In [99]:
print(model(X, 0.013000000000000001, [1, 0.5], 1755))

1660 [47.23214285  3.91071429] 45.937499999999986
1678 [47.23214285  3.91071429] 45.937499999999986
1688 [47.23214285  3.91071429] 45.937499999999986
1690 [47.23214285  3.91071429] 45.937499999999986
1701 [47.23214285  3.91071429] 45.937499999999986
1707 [47.23214285  3.91071429] 45.937499999999986
1709 [47.23214285  3.91071429] 45.937499999999986
1712 [47.23214285  3.91071429] 45.937499999999986
1713 [47.23214285  3.91071429] 45.937499999999986
1724 [47.23214285  3.91071429] 45.937499999999986
1728 [47.23214285  3.91071429] 45.937499999999986
1729 [47.23214285  3.91071429] 45.937499999999986
1731 [47.23214285  3.91071429] 45.937499999999986
1732 [47.23214285  3.91071429] 45.937499999999986
1740 [47.23214285  3.91071429] 45.937499999999986
1749 [47.23214286  3.91071429] 45.937499999999986
None


## Alpha подобрана тем же методом перебора.
## Ответ: alpha = 0.013000000000000001, iter = 1749